File: NesneOdev/NesneOdev.py
import random
from abc import ABC, abstractmethod

# 1. ÖZEL HATA YÖNETİMİ
class KuantumCokusuException(Exception):
    def __init__(self, nesne_id):
        # Hata mesajı formatı
        super().__init__(f"SİSTEM ÇÖKTÜ! TAHLİYE BAŞLATILIYOR... (Patlayan Nesne ID: {nesne_id})")

# 2. ARAYÜZ (INTERFACE) - Python'da Abstract Class olarak tanımlanır
class IKritik(ABC):
    @abstractmethod
    def acil_durum_sogutmasi(self): #
        pass

# 3. TEMEL YAPI (ABSTRACT CLASS)
class KuantumNesnesi(ABC):
    def __init__(self, nesne_id, tehlike_seviyesi):
        self.id = nesne_id #
        self.tehlike_seviyesi = tehlike_seviyesi #
        self._stabilite = 100.0 # Varsayılan başlangıç

    # Kapsülleme (Encapsulation) - Property Decorator
    @property
    def stabilite(self):
        return self._stabilite

    @stabilite.setter
    def stabilite(self, value):
        # 0-100 kontrolü
        if value <= 0:
            self._stabilite = 0
            # Hata fırlatma
            raise KuantumCokusuException(self.id)
        elif value > 100:
            self._stabilite = 100.0
        else:
            self._stabilite = value

    @abstractmethod
    def analiz_et(self): #
        pass

    def durum_bilgisi(self): #
        return f"ID: {self.id} | Stabilite: %{self.stabilite:.2f} | Tehlike: {self.tehlike_seviyesi}"

# 4. NESNE ÇEŞİTLERİ (INHERITANCE)

# A. VeriPaketi
class VeriPaketi(KuantumNesnesi):
    def __init__(self, nesne_id):
        super().__init__(nesne_id, 1)

    def analiz_et(self):
        self.stabilite -= 5 #
        print("Veri içeriği okundu.")

# B. KaranlikMadde - Hem KuantumNesnesi hem IKritik (Çoklu Kalıtım)
class KaranlikMadde(KuantumNesnesi, IKritik):
    def __init__(self, nesne_id):
        super().__init__(nesne_id, 5)

    def analiz_et(self):
        self.stabilite -= 15 #

    def acil_durum_sogutmasi(self):
        self.stabilite += 50 #
        print(f"{self.id} soğutuldu. Stabilite yenilendi.")

# C. AntiMadde
class AntiMadde(KuantumNesnesi, IKritik):
    def __init__(self, nesne_id):
        super().__init__(nesne_id, 10)

    def analiz_et(self):
        self.stabilite -= 25 #
        print("Evrenin dokusu titriyor...") #

    def acil_durum_sogutmasi(self):
        self.stabilite += 50
        print(f"{self.id} üzerinde kritik soğutma yapıldı!")

# 5. OYNANIŞ DÖNGÜSÜ (MAIN)
def main():
    envanter = [] # Generic List karşılığı
    sayac = 1
    
    print("Omega Sektörü Kuantum Veri Ambarı'na Hoşgeldiniz...")

    while True: # Sonsuz döngü
        try:
            print("\n--- KUANTUM AMBARI KONTROL PANELİ ---")
            print("1. Yeni Nesne Ekle")
            print("2. Tüm Envanteri Listele")
            print("3. Nesneyi Analiz Et")
            print("4. Acil Durum Soğutması Yap")
            print("5. Çıkış")
            secim = input("Seçiminiz: ")

            if secim == "1":
                tur = random.randint(1, 3)
                yeni_id = f"NESNE-{sayac}"
                sayac += 1
                yeni_nesne = None

                if tur == 1: yeni_nesne = VeriPaketi(yeni_id)
                elif tur == 2: yeni_nesne = KaranlikMadde(yeni_id)
                elif tur == 3: yeni_nesne = AntiMadde(yeni_id)
                
                envanter.append(yeni_nesne)
                print(f"{type(yeni_nesne).__name__} eklendi: {yeni_id}")

            elif secim == "2":
                print("\n--- ENVANTER DURUMU ---")
                for nesne in envanter:
                    print(nesne.durum_bilgisi()) # Polimorfizm

            elif secim == "3":
                inp_id = input("Analiz edilecek ID: ")
                # List comprehension ile arama
                bulunan = next((x for x in envanter if x.id == inp_id), None)
                if bulunan:
                    bulunan.analiz_et()
                    print(f"Güncel Stabilite: {bulunan.stabilite}")
                else:
                    print("Nesne bulunamadı!")

            elif secim == "4":
                inp_id = input("Soğutulacak ID: ")
                bulunan = next((x for x in envanter if x.id == inp_id), None)
                if bulunan:
                    # Type Checking (isinstance)
                    if isinstance(bulunan, IKritik):
                        bulunan.acil_durum_sogutmasi()
                    else:
                        print("Bu nesne soğutulamaz!") #
                else:
                    print("Nesne bulunamadı!")

            elif secim == "5":
                print("Çıkış yapılıyor...")
                break

        except KuantumCokusuException as ex: #
            print("\n**************************************")
            print(str(ex).upper())
            print("**************************************")
            break # Game over
        except Exception as ex:
            print(f"Hata: {ex}")

if __name__ == "__main__":
    main()

File: NesneOdev/NesneOdev.js
const readline = require('readline');

// Konsol Okuma Arayüzü
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Yardımcı Fonksiyon: Input almak için
const soruSor = (soru) => {
    return new Promise((resolve) => rl.question(soru, resolve));
};

// 1. ÖZEL HATA YÖNETİMİ
class KuantumCokusuException extends Error {
    constructor(id) {
        super(`SİSTEM ÇÖKTÜ! TAHLİYE BAŞLATILIYOR... (Patlayan Nesne ID: ${id})`); // [cite: 44]
        this.name = "KuantumCokusuException";
    }
}

// 2. ARAYÜZ SİMÜLASYONU
// JS'de Interface yoktur, "Type Checking" için bu sınıfı miras almaları gerekecek.
// Ancak JS çoklu kalıtımı desteklemez.
// Bu yüzden "Mixin" veya "Duck Typing" kullanacağız.
// Ödevde "is/as" kontrolü istendiği için, nesnenin metodu var mı diye bakacağız veya
// manuel bir etiket sistemi kullanacağız. En akademik yöntem:
class IKritik {
    acilDurumSogutmasi() {
        throw new Error("Bu metod implement edilmeli!");
    }
}

// 3. TEMEL YAPI (ABSTRACT CLASS SİMÜLASYONU)
class KuantumNesnesi {
    constructor(id, tehlikeSeviyesi) {
        if (this.constructor === KuantumNesnesi) {
            throw new Error("Abstract sınıf nesneleştirilemez!"); // [cite: 16]
        }
        this.id = id;
        this.tehlikeSeviyesi = tehlikeSeviyesi;
        this._stabilite = 100.0;
    }

    // Kapsülleme
    get stabilite() {
        return this._stabilite;
    }

    set stabilite(value) {
        if (value <= 0) {
            this._stabilite = 0;
            throw new KuantumCokusuException(this.id); // [cite: 44]
        } else if (value > 100) {
            this._stabilite = 100;
        } else {
            this._stabilite = value;
        }
    }

    analizEt() {
        throw new Error("Bu metod override edilmeli!"); // Abstract Metot
    }

    durumBilgisi() {
        return `ID: ${this.id} | Stabilite: %${this._stabilite.toFixed(2)} | Tehlike: ${this.tehlikeSeviyesi}`;
    }
}

// 4. NESNE ÇEŞİTLERİ

class VeriPaketi extends KuantumNesnesi {
    constructor(id) {
        super(id, 1);
    }

    analizEt() {
        this.stabilite -= 5; // [cite: 32]
        console.log("Veri içeriği okundu.");
    }
}

// IKritik davranışını simüle eden sınıflar
class KaranlikMadde extends KuantumNesnesi {
    constructor(id) {
        super(id, 5);
    }

    // IKritik Implementasyonu
    acilDurumSogutmasi() {
        this.stabilite += 50; // [cite: 36]
        console.log(`${this.id} soğutuldu. Stabilite yenilendi.`);
    }

    analizEt() {
        this.stabilite -= 15; // [cite: 35]
    }
}

class AntiMadde extends KuantumNesnesi {
    constructor(id) {
        super(id, 10);
    }

    // IKritik Implementasyonu
    acilDurumSogutmasi() {
        this.stabilite += 50;
        console.log(`${this.id} üzerinde kritik soğutma yapıldı!`);
    }

    analizEt() {
        this.stabilite -= 25; // [cite: 39]
        console.log("Evrenin dokusu titriyor..."); // [cite: 40]
    }
}

// 5. MAIN DÖNGÜSÜ
async function main() {
    let envanter = [];
    let sayac = 1;
    console.log("Omega Sektörü Kuantum Veri Ambarı'na Hoşgeldiniz...");

    while (true) { // [cite: 46]
        try {
            console.log("\n--- KUANTUM AMBARI KONTROL PANELİ ---");
            console.log("1. Yeni Nesne Ekle");
            console.log("2. Tüm Envanteri Listele");
            console.log("3. Nesneyi Analiz Et");
            console.log("4. Acil Durum Soğutması Yap");
            console.log("5. Çıkış");
            
            const secim = await soruSor("Seçiminiz: "); // Async input

            if (secim === "1") {
                const tur = Math.floor(Math.random() * 3) + 1;
                const yeniId = `NESNE-${sayac++}`;
                let yeniNesne;

                if (tur === 1) yeniNesne = new VeriPaketi(yeniId);
                else if (tur === 2) yeniNesne = new KaranlikMadde(yeniId);
                else if (tur === 3) yeniNesne = new AntiMadde(yeniId);

                envanter.push(yeniNesne);
                console.log(`${yeniNesne.constructor.name} eklendi: ${yeniId}`);
            }
            else if (secim === "2") {
                console.log("\n--- ENVANTER DURUMU ---");
                envanter.forEach(nesne => {
                    console.log(nesne.durumBilgisi());
                });
            }
            else if (secim === "3") {
                const id = await soruSor("Analiz edilecek ID: ");
                const nesne = envanter.find(n => n.id === id);
                
                if (nesne) {
                    nesne.analizEt();
                    console.log(`Güncel Stabilite: ${nesne.stabilite}`);
                } else {
                    console.log("Nesne bulunamadı!");
                }
            }
            else if (secim === "4") {
                const id = await soruSor("Soğutulacak ID: ");
                const nesne = envanter.find(n => n.id === id);

                if (nesne) {
                    // Type Checking: JS'de 'interface' olmadığı için metod kontrolü veya instance kontrolü yapılır.
                    // [cite: 57] "IKritik olup olmadığı kontrolü"
                    if (nesne instanceof KaranlikMadde || nesne instanceof AntiMadde) {
                        nesne.acilDurumSogutmasi();
                    } else {
                        console.log("Bu nesne soğutulamaz!"); // [cite: 58]
                    }
                } else {
                    console.log("Nesne bulunamadı!");
                }
            }
            else if (secim === "5") {
                console.log("Çıkış yapılıyor...");
                rl.close();
                break;
            }

        } catch (e) {
            if (e.name === "KuantumCokusuException") { // [cite: 59]
                console.log("\n**************************************");
                console.log(e.message.toUpperCase());
                console.log("**************************************");
                rl.close();
                process.exit(1); // Game Over
            } else {
                console.log(`Beklenmedik Hata: ${e.message}`);
            }
        }
    }
}

main();